### ISSUE TITLE
functools lru_cache causing memory build up?

### ISSUE THREAD

**Issue by STBadman** [dstansby/pfsspy#148]
![functools lru_cache](https://user-images.githubusercontent.com/54911235/76153446-ead30600-6080-11ea-94c8-9c7aa1706e35.png)

I'm finding for applications where I am running pfsspy in a loop, e.g. to iterate over many days of magnetograms, my memory is filling up steadily and eventually crashes. 

I've found at least one repeatable test case where I generate a pfsspy.Output object and then access it's contents. The attached image shows some code I run on the left and my system monitor on the right. I first run a loop where I create the Output object, po, and assign the bc property to a new variable and repeat this 4 times, always overwriting the same variables. In the system monitor you can see 4 pulse of CPU activity, the first time the memory usage increases, the remaining 3 it stays more or less constant, which is expected.

However, I next do the same loop but instead of accessing bc, I access bg. This is the next 4 pulses of cpu activity and you can see the memory usage climbs each time even though the variables are being overwritten in the same way.

Looking at the pfsspy.Output source code, the difference seems to be that bg has a functools.lru_cache decorator. I don't fully understand what this does but it seems to be causing this memory "leakage". Is there a solution to prevent this? What is the function of this decorator?

**Comment by dstansby** [comment]
[`lru_cache`](https://docs.python.org/3/library/functools.html#functools.lru_cache) caches the output of a function, such that if the function is called multiple times the output doesn't have to be computed each time.

I wonder if this is doing what is intended. If you print `hex(id(bg))` (this prints the memory address of `bg`) does it print the same value each time or does it change?

**Comment by STBadman** [comment]
Memory address  output from 

`>>> sys.stdout.write(ii, ": ", hex(id(bg),"\n") `

Is :

0 :  0x7faec481df30
1 :  0x7faec481dee0
2 :  0x7faec481df80
3 :  0x7faec481de90

So it changes with each iteration even though I'm overwriting the variable bg each time.  Does this therefore mean each of the above memory addresses is storing a different instance of the bg value? 

(Out of curiosity I also ran :

```
>>> for ii in range(100):
>>>    arr=np.ones([1000,1000])
>>>    print(ii,": ",hex(id(arr)))
```

which gave output :

0 :  0x7faec33e3800
1 :  0x7faec17f2940
2 :  0x7faec33e3800
3 :  0x7faec17f2940
4 :  0x7faec33e3800
5 :  0x7faec17f2940
....

so in this case, every other run uses the same memory address, is this standard behaviour?)

**Comment by STBadman** [comment]
One solution appears to be to restrict the lru_cache call to have maxsize=1 (the default is 128) 
https://docs.python.org/3/library/functools.html#functools.lru_cache

This keeps my memory usage constant (the memory location of bg still changes each call) . Here's a similar screenshot to before, I ran 8 times without memory usage showing much change.

![functools lru_cache_maxsize](https://user-images.githubusercontent.com/54911235/76174476-b0db3000-6164-11ea-9cec-e14bc883baf3.png)

I think I understand that the maxsize allows caching different calls of the same function up to some number of instances, so I think this might be an acceptable solution since for any given pfsspy.Output object, there is only one version of bg (one set of inputs and therefore the same output) and so you don't need the possibility to cache multiple versions of the same bg . Do you agree with that?

**Comment by dstansby** [comment]
That sounds sensible. Would you be happy trying to put in a pull request with that change? There isn't a guide in the pfsspy docs, but there's some good semi-generic instructions here: https://matplotlib.org/devel/contributing.html#contributing-code

**Comment by STBadman** [comment]
Sure, will give that a go!

**Comment by STBadman** [comment]
I submitted the pull request above. However one further thing to add.

I read a bit more the functools docs and found the [cached_property](https://docs.python.org/3/library/functools.html#functools.cached_property) decorator which I think is more appropriate for bg. 'cached_property' is meant to be used to "Transform a method of a class into a property whose value is computed once and then cached as a normal attribute for the life of the instance." which I think is exactly what you are intending for bg.

 functools.lru_cached does seem to have known issues working with class properties (see [https://github.com/python/mypy/issues/5858](https://github.com/python/mypy/issues/5858). For example for a given Output instance, Output.bg does not appear to inherit the lru_cache associated functions such as cache_info() and cache_clear() which are meant to allow control of the cache after it is instantiated. 

However, the issue with functools.cached_property is that it is new to python 3.8. I'm not sure if you want to require users to have python 3.8?

**Comment by dstansby** [comment]
Fixed by #150
